internal/filesystem/cgofuse/table.go

```diff
@@ -1,140 +1,147 @@
 //go:build !nofuse
-// +build !nofuse
 
 package cgofuse
 
 import (
 	"errors"
 	"fmt"
 	"io/fs"
+	"math"
 	"sync"
-)
-
-// TODO: review whole file; quickly ported
 
-type (
-	handle     = uint64
-	fileHandle struct {
-		goFile fs.File
-		ioMu   sync.Mutex // TODO: name and responsibility; currently applies to the position cursor
-	}
-	fMap map[handle]*fileHandle
+	fserrors "github.com/djdv/go-filesystem-utils/internal/filesystem/errors"
+	"github.com/djdv/go-filesystem-utils/internal/generic"
+	"github.com/winfsp/cgofuse/fuse"
 )
 
-/*
-type (
-	handle       = uint64
-	fileResource struct {
-		sync.Mutex
-		fs.File
-	}
-	fMap map[handle]fileResource
+type fileTable struct {
+	sync.RWMutex
+	files []*fileHandle
+}
 
-	directoryResource struct {
-		sync.Mutex
-		// TODO: copy the note about implementation differences here
-		// Fuse only implies this context has to be valid for open and creation operations
-		// and thus some implementations return zero or garbage values
-		// during operations like readdir (where we might need them depending on the platform)
-		fuseContext struct {
-			uid, gid uint32
-			// pid omitted as it's currently unused
-		}
-		fs.ReadDirFile
-	}
-	dMap map[handle]directoryResource
+const (
+	errorHandle = math.MaxUint64
+	// TODO: handleMax needs to be configurable.
+	// This value is arbitrary.
+	handleMax = 2048
 )
-*/
 
 var (
 	errInvalidHandle = errors.New("handle not found")
 	errFull          = errors.New("all slots filled")
 )
 
-func newFileTable() fileTable { return &fileTableStruct{files: make(fMap)} }
-
-type (
-	fileTable interface {
-		Add(fs.File) (handle, error)
-		Exists(handle) bool
-		Get(handle) (*fileHandle, error)
-		Remove(handle) error
-		Length() int
-		Close() error
-	}
-	fileTableStruct struct {
-		sync.RWMutex
-		index   uint64
-		wrapped bool // if true; we start reclaiming dead index values
-		files   fMap
-	}
-)
+func newFileTable() *fileTable { return &fileTable{files: make([]*fileHandle, 0)} }
 
-func (ft *fileTableStruct) Add(f fs.File) (handle, error) {
+func (ft *fileTable) add(f fs.File) (fileDescriptor, error) {
```

Comment:
The logic for growing the table is wrong.
If we don't have enough space to store an entry, we grow the slice.
But on the next call, we don't actually check the capacity and extend the slice + insert the element. We just grow the slice again, with +1 length, and insert into it.
Will fix.

Reply:
Resolved by https://github.com/djdv/go-filesystem-utils/pull/26/commits/09b93ff5683c515dc202115584017387b1c237cb